From 3651da717bd3ff9ff7009913fb9dcb7727a11df2 Mon Sep 17 00:00:00 2001
From: Viktor Rosendahl <viktor.rosendahl@gmail.com>
Date: Tue, 28 Jan 2020 00:28:23 +0100
Subject: [PATCH] traceshark.pro: Add option to use the system qcustomplot

commit 3651da717bd3ff9ff7009913fb9dcb7727a11df2 upstream

This new USE_SYSTEM_QCUSTOMPLOT makes it possible to use a QCustomPlot
library that is already installed by the distribution. Using this
option right now is probably a bad idea because the internal
QCustomPLot library contains important performance hacks.

Signed-off-by: Viktor Rosendahl <viktor.rosendahl@gmail.com>
---
 traceshark.pro      | 22 ++++++++++++++-
 ui/cursor.h         |  4 +--
 ui/mainwindow.cpp   |  2 +-
 ui/migrationarrow.h |  4 +--
 ui/migrationline.h  |  4 +--
 ui/qcustomplot.h    | 66 +++++++++++++++++++++++++++++++++++++++++++++
 ui/tasktoolbar.cpp  |  4 +--
 ui/traceplot.h      |  4 +--
 ui/yaxisticker.h    |  4 +--
 9 files changed, 100 insertions(+), 14 deletions(-)
 create mode 100644 ui/qcustomplot.h

diff --git a/traceshark.pro b/traceshark.pro
index 8868fce..e594b60 100644
--- a/traceshark.pro
+++ b/traceshark.pro
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 #
 #  Traceshark - a visualizer for visualizing ftrace and perf traces
-#  Copyright (C) 2014-2019  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+#  Copyright (C) 2014-2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
 #
 # This file is dual licensed: you can use it either under the terms of
 # the GPL, or the BSD license, at your option.
@@ -247,6 +247,16 @@
 # always have the scheduling graphs drawn with a width of 1.
 # DISABLE_OPENGL = yes
 
+# Uncomment this to use the libqcustomplot of your system. At the time of
+# writing, this is a bad idea. Only do this if you know exactly what you are
+# doing. traceshark has its own QCustomPlot, which contains important
+# performance fixes that are unlikely to be in the system libqcustomplot.
+# On the other hand, if you do not care about performance and have at least one
+# other application that uses libqcustomplot and you want to save about 1.3 MB
+# of disk space, then this is for you. If libqcustomplot was compiled without
+# OpenGL support, then you probably want to disable OpenGL above as well.
+# USE_SYSTEM_QCUSTOMPLOT = yes
+
 # Uncomment this for debug symbols
 # USE_DEBUG_FLAG = -g
 
@@ -284,10 +294,12 @@
 # Header files
 #
 
+!equals(USE_SYSTEM_QCUSTOMPLOT, yes) {
 HEADERS       = qcustomplot/qcustomplot.h
 HEADERS      += qcustomplot/qcppointer.h
 HEADERS      += qcustomplot/qcppointer_impl.h
 HEADERS      += qcustomplot/qcplist.h
+}
 
 HEADERS      +=  ui/abstracttaskmodel.h
 HEADERS      +=  ui/cpuselectdialog.h
@@ -306,6 +318,7 @@ HEADERS      +=  ui/licensedialog.h
 HEADERS      +=  ui/mainwindow.h
 HEADERS      +=  ui/migrationarrow.h
 HEADERS      +=  ui/migrationline.h
+HEADERS      +=  ui/qcustomplot.h
 HEADERS      +=  ui/statslimitedmodel.h
 HEADERS      +=  ui/statsmodel.h
 HEADERS      +=  ui/tableview.h
@@ -383,7 +396,9 @@ HEADERS      +=  vtl/time.h
 # Source files
 #
 
+!equals(USE_SYSTEM_QCUSTOMPLOT, yes) {
 SOURCES       = qcustomplot/qcustomplot.cpp
+}
 
 SOURCES      +=  ui/abstracttaskmodel.cpp
 SOURCES      +=  ui/cpuselectdialog.cpp
@@ -494,6 +509,11 @@ equals(USE_HARDENING_CXXFLAGS, yes) {
 OUR_FLAGS += $${HARDENING_CXXFLAGS}
 }
 
+equals(USE_SYSTEM_QCUSTOMPLOT, yes) {
+OUR_FLAGS += -DCONFIG_SYSTEM_QCUSTOMPLOT
+LIBS += -lqcustomplot
+}
+
 OUR_NORMAL_CXXFLAGS = -pedantic -Wall -std=c++11
 OUR_NORMAL_CFLAGS = -pedantic -Wall -std=c11
 
diff --git a/ui/cursor.h b/ui/cursor.h
index a6fcce6..38cafef 100644
--- a/ui/cursor.h
+++ b/ui/cursor.h
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2015-2019  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2015-2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -56,7 +56,7 @@
 #include <QColor>
 
 #include "misc/traceshark.h"
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 #include "vtl/time.h"
 
 class Cursor : public QCPItemLine
diff --git a/ui/mainwindow.cpp b/ui/mainwindow.cpp
index b80a83c..0d13822 100644
--- a/ui/mainwindow.cpp
+++ b/ui/mainwindow.cpp
@@ -83,7 +83,7 @@
 #include "misc/traceshark.h"
 #include "threads/workqueue.h"
 #include "threads/workitem.h"
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 #include "vtl/compiler.h"
 #include "vtl/error.h"
 
diff --git a/ui/migrationarrow.h b/ui/migrationarrow.h
index 2de834b..7d5ff0b 100644
--- a/ui/migrationarrow.h
+++ b/ui/migrationarrow.h
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2015, 2017  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2015, 2017, 2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -54,7 +54,7 @@
 #define MIGRATIONARROW_H
 
 #include <QColor>
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 
 class MigrationArrow : public QCPItemLine
 {
diff --git a/ui/migrationline.h b/ui/migrationline.h
index 8e33060..4763292 100644
--- a/ui/migrationline.h
+++ b/ui/migrationline.h
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2015, 2017  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2015, 2017, 2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -53,7 +53,7 @@
 #ifndef MIGRATIONLINE_H
 #define MIGRATIONLINE_H
 
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 
 class MigrationLine : public QCPItemLine
 {
diff --git a/ui/qcustomplot.h b/ui/qcustomplot.h
new file mode 100644
index 0000000..68226be
--- /dev/null
+++ b/ui/qcustomplot.h
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
+/*
+ * Traceshark - a visualizer for visualizing ftrace and perf traces
+ * Copyright (C) 2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ *
+ * This file is dual licensed: you can use it either under the terms of
+ * the GPL, or the BSD license, at your option.
+ *
+ *  a) This program is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public
+ *     License along with this library; if not, write to the Free
+ *     Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ *     MA 02110-1301 USA
+ *
+ * Alternatively,
+ *
+ *  b) Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *     1. Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *     2. Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *     CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *     INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *     MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *     DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *     SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ *     NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *     LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ *     HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *     CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ *     OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ *     EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef UI_QCUSTOMPLOT_H
+#define UI_QCUSTOMPLOT_H
+
+#ifdef CONFIG_SYSTEM_QCUSTOMPLOT
+
+#include <qcustomplot.h>
+
+#else /* CONFIG_SYSTEM_QCUSTOMPLOT */
+
+#include "qcustomplot/qcustomplot.h"
+
+#endif
+
+#endif
diff --git a/ui/tasktoolbar.cpp b/ui/tasktoolbar.cpp
index 8ab1696..4c879e0 100644
--- a/ui/tasktoolbar.cpp
+++ b/ui/tasktoolbar.cpp
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2015-2019  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2015-2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -67,7 +67,7 @@
 #include "misc/maplist.h"
 #include "misc/resources.h"
 #include "misc/traceshark.h"
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 #include "vtl/error.h"
 
 #define DEFINE_PIDMAP_ITERATOR(name) \
diff --git a/ui/traceplot.h b/ui/traceplot.h
index 1743af7..86457b8 100644
--- a/ui/traceplot.h
+++ b/ui/traceplot.h
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2015, 2017  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2015, 2017, 2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -53,7 +53,7 @@
 #ifndef TRACEPLOT_H
 #define TRACEPLOT_H
 
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 
 class TracePlot : public QCustomPlot
 {
diff --git a/ui/yaxisticker.h b/ui/yaxisticker.h
index a5e1a6a..1e1895c 100644
--- a/ui/yaxisticker.h
+++ b/ui/yaxisticker.h
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0-or-later OR BSD-2-Clause)
 /*
  * Traceshark - a visualizer for visualizing ftrace and perf traces
- * Copyright (C) 2017  Viktor Rosendahl <viktor.rosendahl@gmail.com>
+ * Copyright (C) 2017, 2020  Viktor Rosendahl <viktor.rosendahl@gmail.com>
  *
  * This file is dual licensed: you can use it either under the terms of
  * the GPL, or the BSD license, at your option.
@@ -55,7 +55,7 @@
 
 #include <QVector>
 #include <QString>
-#include "qcustomplot/qcustomplot.h"
+#include "ui/qcustomplot.h"
 
 class YAxisTicker : public QCPAxisTicker {
 public:
-- 
2.20.1

